make-pom does not translate ivy revision ranges to valid maven revision ranges

Discovered this when trying to download bijection-guava via this pom: view-source:http://search.maven.org/remotecontent?filepath=com/twitter/bijection-guava_2.9.3/0.6.0/bijection-guava_2.9.3-0.6.0.pom

``` xml
<dependency>
<groupId>com.google.code.findbugs</groupId>
<artifactId>jsr305</artifactId>
<version>1.3.+</version>
</dependency>
```

Isolated, the failure looks like so:

``` console
$ cat pom.xml 
<?xml version='1.0' encoding='UTF-8'?>
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0">
    <modelVersion>4.0.0</modelVersion>
    <groupId>fake</groupId>
    <artifactId>fake</artifactId>
    <packaging>jar</packaging>
    <version>0.0.1</version>
    <dependencies>
        <dependency>
            <groupId>com.twitter</groupId>
            <artifactId>bijection-guava_2.9.3</artifactId>
            <version>0.6.0</version>
        </dependency>
    </dependencies>
</project>
$ mvn -U dependency:tree
[INFO] Scanning for projects...
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building fake 0.0.1
[INFO] ------------------------------------------------------------------------
Downloading: http://repo.maven.apache.org/maven2/com/twitter/bijection-guava_2.9.3/0.6.0/bijection-guava_2.9.3-0.6.0.pom
Downloaded: http://repo.maven.apache.org/maven2/com/twitter/bijection-guava_2.9.3/0.6.0/bijection-guava_2.9.3-0.6.0.pom (0 B at 0.0 KB/sec)
Downloading: http://repo.maven.apache.org/maven2/com/twitter/bijection-core_2.9.3/0.6.0/bijection-core_2.9.3-0.6.0.pom
Downloaded: http://repo.maven.apache.org/maven2/com/twitter/bijection-core_2.9.3/0.6.0/bijection-core_2.9.3-0.6.0.pom (0 B at 0.0 KB/sec)
Downloading: http://repo.maven.apache.org/maven2/org/scala-lang/scala-library/2.9.3/scala-library-2.9.3.pom
Downloaded: http://repo.maven.apache.org/maven2/org/scala-lang/scala-library/2.9.3/scala-library-2.9.3.pom (0 B at 0.0 KB/sec)
Downloading: http://repo.maven.apache.org/maven2/com/google/code/findbugs/jsr305/1.3.+/jsr305-1.3.+.pom
[WARNING] The POM for com.google.code.findbugs:jsr305:jar:1.3.+ is missing, no dependency information available
Downloading: http://repo.maven.apache.org/maven2/com/google/guava/guava/14.0/guava-14.0.pom
Downloaded: http://repo.maven.apache.org/maven2/com/google/guava/guava/14.0/guava-14.0.pom (0 B at 0.0 KB/sec)
Downloading: http://repo.maven.apache.org/maven2/com/google/guava/guava-parent/14.0/guava-parent-14.0.pom
Downloaded: http://repo.maven.apache.org/maven2/com/google/guava/guava-parent/14.0/guava-parent-14.0.pom (0 B at 0.0 KB/sec)
Downloading: http://repo.maven.apache.org/maven2/com/twitter/bijection-guava_2.9.3/0.6.0/bijection-guava_2.9.3-0.6.0.jar
Downloading: http://repo.maven.apache.org/maven2/com/twitter/bijection-core_2.9.3/0.6.0/bijection-core_2.9.3-0.6.0.jar
Downloading: http://repo.maven.apache.org/maven2/org/scala-lang/scala-library/2.9.3/scala-library-2.9.3.jar
Downloading: http://repo.maven.apache.org/maven2/com/google/guava/guava/14.0/guava-14.0.jar
Downloading: http://repo.maven.apache.org/maven2/com/google/code/findbugs/jsr305/1.3.+/jsr305-1.3.+.jar
Downloaded: http://repo.maven.apache.org/maven2/com/twitter/bijection-guava_2.9.3/0.6.0/bijection-guava_2.9.3-0.6.0.jar (0 B at 0.0 KB/sec)
Downloaded: http://repo.maven.apache.org/maven2/com/google/guava/guava/14.0/guava-14.0.jar (0 B at 0.0 KB/sec)
Downloaded: http://repo.maven.apache.org/maven2/org/scala-lang/scala-library/2.9.3/scala-library-2.9.3.jar (0 B at 0.0 KB/sec)
Downloaded: http://repo.maven.apache.org/maven2/com/twitter/bijection-core_2.9.3/0.6.0/bijection-core_2.9.3-0.6.0.jar (0 B at 0.0 KB/sec)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 2.645s
[INFO] Finished at: Tue Feb 25 08:10:33 MST 2014
[INFO] Final Memory: 8M/150M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project fake: Could not resolve dependencies for project fake:fake:jar:0.0.1: Could not find artifact com.google.code.findbugs:jsr305:jar:1.3.+ in central (http://repo.maven.apache.org/maven2) -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException

```

The corresponding sbt build snippet that leads to this pom is:

``` scala
    libraryDependencies ++= Seq(
      "com.google.code.findbugs" % "jsr305" % "1.3.+",
      "com.google.guava" % "guava" % "14.0"
    )
```

Peeking at MakePom does show a straight mapping with no treatment to ensure things like ivy's support for a trailing + get mapped to maven [,) for example which is valid in maven poms: https://github.com/sbt/sbt/blob/b7c50d42698f8309c76ed5fd0efbb3ef61ee601e/ivy/src/main/scala/sbt/MakePom.scala

See this workaround pull request in bijection-guava for reference: https://github.com/twitter/bijection/pull/164

Confirming that + -> [,) works as a potential solution here:

Before:

``` console
$ cat pom.xml 

<?xml version='1.0' encoding='UTF-8'?>
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0">
    <modelVersion>4.0.0</modelVersion>
    <groupId>fake</groupId>
    <artifactId>fake</artifactId>
    <packaging>jar</packaging>
    <version>0.0.1</version>
    <dependencies>
        <dependency>
            <groupId>com.google.code.findbugs</groupId>
            <artifactId>jsr305</artifactId>
            <version>1.3.+</version>
        </dependency>
    </dependencies>
</project>
$ mvn -U dependency:tree
[INFO] Scanning for projects...
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building fake 0.0.1
[INFO] ------------------------------------------------------------------------
Downloading: http://repo.maven.apache.org/maven2/com/google/code/findbugs/jsr305/1.3.+/jsr305-1.3.+.pom
[WARNING] The POM for com.google.code.findbugs:jsr305:jar:1.3.+ is missing, no dependency information available
Downloading: http://repo.maven.apache.org/maven2/com/google/code/findbugs/jsr305/1.3.+/jsr305-1.3.+.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 1.128s
[INFO] Finished at: Tue Feb 25 08:22:10 MST 2014
[INFO] Final Memory: 8M/150M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project fake: Could not resolve dependencies for project fake:fake:jar:0.0.1: Could not find artifact com.google.code.findbugs:jsr305:jar:1.3.+ in central (http://repo.maven.apache.org/maven2) -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
```

After:

``` console
$ cat pom.xml 

<?xml version='1.0' encoding='UTF-8'?>
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0">
    <modelVersion>4.0.0</modelVersion>
    <groupId>fake</groupId>
    <artifactId>fake</artifactId>
    <packaging>jar</packaging>
    <version>0.0.1</version>
    <dependencies>
        <dependency>
            <groupId>com.google.code.findbugs</groupId>
            <artifactId>jsr305</artifactId>
            <version>[1.3,)</version>
        </dependency>
    </dependencies>
</project>
$ mvn -U dependency:tree
[INFO] Scanning for projects...
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building fake 0.0.1
[INFO] ------------------------------------------------------------------------
Downloading: http://repo.maven.apache.org/maven2/com/google/code/findbugs/jsr305/maven-metadata.xml
Downloaded: http://repo.maven.apache.org/maven2/com/google/code/findbugs/jsr305/maven-metadata.xml (368 B at 0.8 KB/sec)
Downloading: http://repo.maven.apache.org/maven2/com/google/code/findbugs/jsr305/2.0.2/jsr305-2.0.2.pom
Downloaded: http://repo.maven.apache.org/maven2/com/google/code/findbugs/jsr305/2.0.2/jsr305-2.0.2.pom (964 B at 6.2 KB/sec)
Downloading: http://repo.maven.apache.org/maven2/com/google/code/findbugs/jsr305/2.0.3/jsr305-2.0.3.pom
Downloaded: http://repo.maven.apache.org/maven2/com/google/code/findbugs/jsr305/2.0.3/jsr305-2.0.3.pom (2 KB at 9.3 KB/sec)
Downloading: http://repo.maven.apache.org/maven2/com/google/code/findbugs/jsr305/2.0.3/jsr305-2.0.3.jar
Downloaded: http://repo.maven.apache.org/maven2/com/google/code/findbugs/jsr305/2.0.3/jsr305-2.0.3.jar (33 KB at 105.8 KB/sec)
[INFO] 
[INFO] --- maven-dependency-plugin:2.1:tree (default-cli) @ fake ---
[INFO] fake:fake:jar:0.0.1
[INFO] \- com.google.code.findbugs:jsr305:jar:2.0.3:compile
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 2.504s
[INFO] Finished at: Tue Feb 25 08:20:07 MST 2014
[INFO] Final Memory: 11M/118M
[INFO] ------------------------------------------------------------------------
```

--- a/ivy/src/main/scala/sbt/MakePom.scala
+++ b/ivy/src/main/scala/sbt/MakePom.scala
@@ -188,13 +188,32 @@ class MakePom(val log: Logger)
 		<dependency>
 			<groupId>{mrid.getOrganisation}</groupId>
 			<artifactId>{mrid.getName}</artifactId>
-			<version>{mrid.getRevision}</version>
+			<version>{makeDependencyVersion(mrid.getRevision)}</version>
 			{ scopeElem(scope) }
 			{ optionalElem(optional) }
 			{ classifierElem(classifier) }
 			{ typeElem(tpe) }
 			{ exclusions(dependency) }
 		</dependency>
+	}
+
+
+
+	def makeDependencyVersion(revision: String): String = {
+		if(revision endsWith "+") try {
+			// TODO - this is the slowest possible implementation.
+			val beforePlus = revision.reverse.dropWhile(_ != '.').drop(1).reverse
+			val lastVersion = beforePlus.reverse.takeWhile(_ != '.').reverse
+			val lastVersionInt = lastVersion.toInt
+			val prefixVersion = beforePlus.reverse.dropWhile(_ != '.').drop(1).reverse
+			s"[$beforePlus, ${prefixVersion}.${lastVersionInt+1})"
+		} catch {
+			case e: NumberFormatException => 
+			  // TODO - if the version deosn't meet our expectations, maybe we just issue a hard
+			  //        error instead of softly ignoring the attempt to rewrite.
+			  //sys.error(s"Could not fix version [$revision] into maven style version")
+			  revision
+		} else revision
 	}
 
 	@deprecated("No longer used and will be removed.", "0.12.1")

--- a/ivy/src/main/scala/sbt/MakePom.scala
+++ b/ivy/src/main/scala/sbt/MakePom.scala
@@ -200,20 +200,39 @@ class MakePom(val log: Logger)
 
 
 	def makeDependencyVersion(revision: String): String = {
-		if(revision endsWith "+") try {
-			// TODO - this is the slowest possible implementation.
-			val beforePlus = revision.reverse.dropWhile(_ != '.').drop(1).reverse
-			val lastVersion = beforePlus.reverse.takeWhile(_ != '.').reverse
+		def plusRange(s:String, shift:Int = 0) = {
+			def pow(i:Int):Int = if (i>0) 10 * pow(i-1) else 1
+			val (prefixVersion, lastVersion) = (s+"0"*shift).reverse.split("\\.",2) match {
+				case Array(revLast,revRest) =>
+					( revRest.reverse + ".", revLast.reverse )
+				case Array(revLast) => ("", revLast.reverse)
+			}
 			val lastVersionInt = lastVersion.toInt
-			val prefixVersion = beforePlus.reverse.dropWhile(_ != '.').drop(1).reverse
-			s"[$beforePlus, ${prefixVersion}.${lastVersionInt+1})"
+			s"[${prefixVersion}${lastVersion},${prefixVersion}${lastVersionInt+pow(shift)})"
+		}
+		val startSym=Set(']','[','(')
+		val stopSym=Set(']','[',')')
+		try {
+			if (revision endsWith ".+") {
+				plusRange(revision.substring(0,revision.length-2))
+			} else if (revision endsWith "+") {
+				val base = revision.take(revision.length-1)
+				// This is a heuristic.  Maven just doesn't support Ivy's notions of 1+, so
+				// we assume version ranges never go beyond 5 siginificant digits.
+				(0 to 5).map(plusRange(base,_)).mkString(",")
+			} else if (startSym(revision(0)) && stopSym(revision(revision.length-1))) {
+				val start = revision(0)
+				val stop = revision(revision.length-1)
+      			val mid = revision.substring(1,revision.length-1)
+      			(if (start == ']') "(" else start) + mid + (if (stop == '[') ")" else stop)
+    		} else revision
 		} catch {
-			case e: NumberFormatException => 
-			  // TODO - if the version deosn't meet our expectations, maybe we just issue a hard
-			  //        error instead of softly ignoring the attempt to rewrite.
-			  //sys.error(s"Could not fix version [$revision] into maven style version")
-			  revision
-		} else revision
+    		case e: NumberFormatException =>
+    		// TODO - if the version doesn't meet our expectations, maybe we just issue a hard
+    		//        error instead of softly ignoring the attempt to rewrite.
+    		//sys.error(s"Could not fix version [$revision] into maven style version")
+    		revision
+		}	
 	}
 
 	@deprecated("No longer used and will be removed.", "0.12.1")

--- a/ivy/src/test/scala/MakePomTest.scala
+++ b/ivy/src/test/scala/MakePomTest.scala
@@ -1,0 +1,29 @@
+package sbt
+
+import java.io.File
+import org.specs2._
+import mutable.Specification
+
+object MakePomTest extends Specification
+{
+	val mp = new MakePom(ConsoleLogger())
+	import mp.{makeDependencyVersion=>v}
+	"MakePom makeDependencyVersion" should {
+		"Handle .+ in versions" in {
+			v("1.+") must_== "[1,2)"
+            v("1.2.3.4.+") must_== "[1.2.3.4,1.2.3.5)"
+			v("12.31.42.+") must_== "[12.31.42,12.31.43)"
+		}
+		/* TODO - do we care about this case?
+         * 1+ --> [1,2),[10,20),[100,200),[1000,2000),[10000,20000),[100000,200000)
+         */
+		"Handle ]* bracket in version ranges" in {
+			v("]1,3]") must_== "(1,3]"
+			v("]1.1,1.3]") must_== "(1.1,1.3]"
+		}
+		"Handle *[ bracket in version ranges" in {
+			v("[1,3[") must_== "[1,3)"
+			v("[1.1,1.3[") must_== "[1.1,1.3)"
+		}
+	}
+}

--- a/sbt/src/sbt-test/dependency-management/make-pom/project/MakePomTest.scala
+++ b/sbt/src/sbt-test/dependency-management/make-pom/project/MakePomTest.scala
@@ -8,8 +8,10 @@ object MakePomTest extends Build
 		readPom <<= makePom map XML.loadFile,
 		TaskKey[Unit]("check-pom") <<= checkPom,
 		TaskKey[Unit]("check-extra") <<= checkExtra,
+		TaskKey[Unit]("check-version-plus-mapping") <<= checkVersionPlusMapping,
 		resolvers += Resolver.sonatypeRepo("snapshots"),
-		makePomConfiguration ~= { _.copy(extra = <extra-tag/>) }
+		makePomConfiguration ~= { _.copy(extra = <extra-tag/>) },
+		libraryDependencies += "com.google.code.findbugs" % "jsr305" % "1.3.+"
 	)
 
 	val readPom = TaskKey[Elem]("read-pom")
@@ -33,6 +35,16 @@ object MakePomTest extends Build
 		if(extra.isEmpty) error("'" + extraTagName + "' not found in generated pom.xml.") else ()
 	}
 	
+	lazy val checkVersionPlusMapping = (readPom) map { (pomXml) =>
+		var found = false
+		for {
+			dep <- pomXml \ "dependencies" \ "dependency"
+			if (dep \ "artifactId").text == "jsr305"
+			if (dep \ "version").text contains "+"
+		} sys.error(s"Found dependency with invalid maven version: $dep")
+		()
+	}
+
 	lazy val checkPom = (readPom, fullResolvers) map { (pomXML, ivyRepositories) =>
 		checkProject(pomXML)
 		withRepositories(pomXML) { repositoriesElement =>

--- a/sbt/src/sbt-test/dependency-management/make-pom/project/MakePomTest.scala
+++ b/sbt/src/sbt-test/dependency-management/make-pom/project/MakePomTest.scala
@@ -40,7 +40,7 @@ object MakePomTest extends Build
 		for {
 			dep <- pomXml \ "dependencies" \ "dependency"
 			if (dep \ "artifactId").text == "jsr305"
-			if (dep \ "version").text contains "+"
+			if (dep \ "version").text != "[1.3, 1.4)"
 		} sys.error(s"Found dependency with invalid maven version: $dep")
 		()
 	}

--- a/sbt/src/sbt-test/dependency-management/make-pom/project/MakePomTest.scala
+++ b/sbt/src/sbt-test/dependency-management/make-pom/project/MakePomTest.scala
@@ -40,7 +40,8 @@ object MakePomTest extends Build
 		for {
 			dep <- pomXml \ "dependencies" \ "dependency"
 			if (dep \ "artifactId").text == "jsr305"
-			if (dep \ "version").text != "[1.3, 1.4)"
+			// TODO - Ignore space here.
+			if (dep \ "version").text != "[1.3,1.4)"
 		} sys.error(s"Found dependency with invalid maven version: $dep")
 		()
 	}

--- a/sbt/src/sbt-test/dependency-management/make-pom/test
+++ b/sbt/src/sbt-test/dependency-management/make-pom/test
@@ -1,2 +1,3 @@
 > check-pom
 > check-extra
+> check-version-plus-mapping

